flowauth/frontend/src/RoleDetails.jsx

```diff
@@ -0,0 +1,320 @@
+      setLifetimeIsValid(false)
+    } else {
+      setLifetimeHelperText("Lifetime must be a number")
+      setLifetimeIsValid(false)
+    }
+  }, [maxLifetime])
+
+  //Make sure entire form is valid on change for any relevant members
+  useEffect(() => {
+    const valid_vars = validation_vars.map(v => typeof(v))
+    if (valid_vars.includes("undefined")){
+      setFormIsValid(false)
+    } else if (nameIsValid && lifetimeIsValid && scopes.length > 0 && members.length > 0 && server_id >= 0){
+      console.log("Form is valid")
+      setFormIsValid(true)
+    } else {
+      setFormIsValid(false)
+    }
+  }, validation_vars)
+
+  const handleNameChange = (event) => {
+    console.log("Name change event handled")
+    setRoleName(event.target.value)
+  }
+
+  const handleLifetimeChange = (event) => {
+    console.log("Lifetime change event handled");
+    setMaxLifetime(event.target.value)
+  }
+
+  const handleServerChange = (event) => {
+    console.log("Server picker event handled");
+    const index = event.target.value
+    setServer(index)
+  }
+
+  const handleMembersChange  = (new_members) => {
+    console.log("Member change event handled")
+    setMembers(new_members.map(x => x.id))
+  }
+
+  const handleScopesChange = (new_scopes) => {
+    console.log("Scopes change event handled")
+    setScopes(new_scopes)
+  }
+
+ //Either update or create a new role on 'submit' button
+ const handleSubmit = async () => {
+  console.log("Role form submitted")
+      // Need to throw an error here if !formIsValid
+    if (!formIsValid) {
+      setIsErrored(true)
+      setErrors(new Error("Uncaught form validation error; please report to Flowminder."))
+    }
+    else if (edit_mode){
+      editRole(
+        role.id,
+        name,
+        scopes.map(s => s.id),
+        members,
+        expiryDate,
+        maxLifetime)
+      .catch((err)=>{
+        setIsErrored(true)
+        setErrors(err)
+      })
+    } else {
+      createRole(
+        name,
+        server_id,
+        scopes.map(s => s.id),
+        members,
+        expiryDate,
+        maxLifetime)
+      .catch((err) => {
+        setIsErrored(true)
+        setErrors(err)
+      })
+    }
+    onClick()
+  };
+
+console.log("Prerendering:")
+console.log("server: ",server_id)
+console.log("Server list: ",JSON.stringify(serverList))
+
+return (
+    <React.Fragment>
+
+      <Grid item xs={12}>
+        <Typography variant="h5" component="h1">
+          {(edit_mode && "Edit Role") || "New Role"}
+        </Typography>
+      </Grid>
+      <Grid item xs={12}>
+        <TextField
+          id="name"
+          label="Name"
+          className={classes.textField}
+          required={true}
+          value={name}
+          onChange={handleNameChange}
+          margin="normal"
+          error={!nameIsValid}
+          helperText={name_helper_text}
+        />
+      </Grid>
+      <MuiPickersUtilsProvider utils={DateFnsUtils}>
```

Comment:
Expiry date defaults to now in CreateRole. It would be preferable for this to default to the server's "latest expiry" (and similarly for the max lifetime to default to the server's max lifetime).

Reply:
I suppose the tricky part here is that the server is chosen within the "CreateRole" view, so isn't known when the view is first opened. In many ways I think it would be cleaner to have a separate "new role" button per server (within RoleList) and remove the server selection from within CreateRole. This could be combined with the "server headings in RoleList" suggestion from my earlier comment. But I appreciate that's not a trivial change to make.

Reply:
`CreateRole` goes and fetches the role list for the appropriate server in current implementation anyway - so I think it should be fine to do.

Reply:
Present in the new `roles` view